src/pkgcheck/checks/metadata.py

```diff
@@ -1044,3 +1044,85 @@ def feed(self, pkg):
             deprecated = set(
                 x for x in bad if x.startswith("no") and x[2:] in self.known_restricts)
             yield BadRestricts(pkg, bad.difference(deprecated), deprecated)
+
+
+class MissingUnpackerDep(base.Warning):
+    """Package is missing dependency on unpacker.
+
+    Package is using an archive format for which unpacker is not provided
+    by the system set, and lacks explicit dependency on the unpacker. """
+
+    __slots__ = ("category", "package", "version", "formats", "unpackers")
+    threshold = base.versioned_feed
+
+    def __init__(self, pkg, formats, unpackers):
+        super().__init__()
+        self._store_cpv(pkg)
+        self.formats = tuple(sorted(formats))
+        self.unpackers = tuple(sorted(unpackers))
+
+    @property
+    def short_desc(self):
+        return (f"Package is using file formats: {', '.join(self.formats)} "
+                f"in SRC_URI but is missing bdepend on unpackers: "
+                f"{' '.join(str(x) for x in self.unpackers)}")
+
+
+class MissingUnpackerDepCheck(base.Template):
+    """Check whether package is missing unpacker dependencies."""
+
+    feed_type = base.versioned_feed
+
+    known_results = (MissingUnpackerDep,)
+    required_addons = (addons.UseAddon,)
+
+    unpackers = {
+        '.zip': ('app-arch/unzip',),
+        '.jar': ('app-arch/unzip',),
+        '.7z': ('app-arch/p7zip',),
+        '.rar': ('app-arch/rar', 'app-arch/unrar'),
+        '.lha': ('app-arch/lha'),
+        '.lzh': ('app-arch/lha'),
+    }
+
+    def __init__(self, options, iuse_handler):
+        super().__init__(options)
+        self.dep_filter = iuse_handler.get_filter()
+        self.fetch_filter = iuse_handler.get_filter('fetchables')
+
+    def feed(self, pkg):
+        # ignore conditionals
+        fetchables, _ = self.fetch_filter((fetchable,), pkg,
+            pkg._get_attr['fetchables'](
+                pkg, allow_missing_checksums=True,
+                ignore_unknown_mirrors=True, skip_default_mirrors=True))
+
+        # formats is used as definitive list of formats for which deps
+        # are missing (since format may allow multiple unpackers)
+        formats = set()
+        # unpackers are used to optimize dep checks
+        unpackers = set()
```

Comment:
I think changing the name of this or the class attr with the same name would be good.